fairseq/criterions/sequence_tagging.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+
+import torch
+import torch.nn.functional as F
+
+from fairseq import metrics, utils
+from fairseq.criterions import FairseqCriterion, register_criterion
+
+
+@register_criterion('sequence_tagging')
+class SequenceTaggingCriterion(FairseqCriterion):
+
+    @staticmethod
+    def add_args(parser):
+        # fmt: off
+        parser.add_argument('--classification-head-name',
+                            default='sentence_classification_head',
+                            help='name of the classification head to use')
+        # fmt: on
+
+    def forward(self, model, sample, reduce=True):
+        """Compute the loss for the given sample.
+
+        Returns a tuple with three elements:
+        1) the loss
+        2) the sample size, which is used as the denominator for the gradient
+        3) logging outputs to display while training
+        """
+        assert (
+            hasattr(model, 'classification_heads')
+            and self.args.classification_head_name in model.classification_heads
+        ), 'model must provide sentence classification head for --criterion=sequence_tagging'
+
+        logits, _ = model(
+            **sample['net_input'],
+            features_only=True,
+            classification_head_name=self.args.classification_head_name,
+        )
+        targets = model.get_targets(sample, [logits]).view(-1)
+        sample_size = sample['ntokens'] - sample['target'].size(0) # number of tokens without eos
+
+        if not self.args.regression_target:
+            logits = logits.view(-1, logits.size(-1))
+            loss = F.nll_loss(
```

Comment:
@myleott Can you please validate this loss calculation?

Reply:
Looks good, but it'd be great if you could test this on a real sequence tagging task first.

Reply:
I added an example with a README 😉 https://github.com/pytorch/fairseq/pull/1709/files#diff-dcc107dc4ee613a610d8174151f0afc4

This made me realize that it might be tricky to use sequence tagging with BPE, since you will have to provide a label for each input BPE token. I clarified that in the README, hope that is enough.

I also added `predict_class` and `predict_tags` convenience methods and used them in the classification and tagging READMEs.

Reply:
By the way, shouldn't I name this task `sentence_tagging` instead of `sequence_tagging`, since we already have `sentence_prediction` and `sentence_ranking`?